eligibility/fixedoracle.go

```diff
@@ -0,0 +1,140 @@
+package eligibility
+
+import (
+	"github.com/spacemeshos/go-spacemesh/log"
+	"sync"
+)
+
+type FixedRolacle struct {
+	honest map[string]struct{}
+	faulty map[string]struct{}
+	emaps  map[uint32]map[string]struct{}
+	mutex  sync.Mutex
+	mapRW  sync.RWMutex
+}
+
+func New() *FixedRolacle {
+	rolacle := &FixedRolacle{}
+	rolacle.honest = make(map[string]struct{})
+	rolacle.faulty = make(map[string]struct{})
+	rolacle.emaps = make(map[uint32]map[string]struct{})
+
+	return rolacle
+}
+
+func (fo *FixedRolacle) update(m map[string]struct{}, client string) {
+	fo.mutex.Lock()
+
+	if _, exist := m[client]; exist {
+		fo.mutex.Unlock()
+		return
+	}
+
+	m[client] = struct{}{}
+
+	fo.mutex.Unlock()
+}
+
+func (fo *FixedRolacle) Register(isHonest bool, client string) {
+	if isHonest {
+		fo.update(fo.honest, client)
+	} else {
+		fo.update(fo.faulty, client)
+	}
+}
+
+func (fo *FixedRolacle) Unregister(isHonest bool, client string) {
+	fo.mutex.Lock()
+	if isHonest {
+		delete(fo.honest, client)
+	} else {
+		delete(fo.faulty, client)
+	}
+	fo.mutex.Unlock()
+}
+
+func cloneMap(m map[string]struct{}) map[string]struct{} {
+	c := make(map[string]struct{}, len(m))
+	for k, v := range m {
```

Comment:
you iterate a map without locking it

Reply:
cloneMap is private and not thread-safe. We never call this function without enforcing safety.